README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# virtualNicheR <img src="man/figures/logo.png" align="right" width="200" />

`virtualNicheR` is an `R` package that generates virtual [fundamental
and realised niches](https://en.wikipedia.org/wiki/Ecological_niche)
that can be used as the basis for virtual ecology experiments examining
questions around ecological niche modelling. The fundamental niches are
defined as [multivariate normal
distributions](https://en.wikipedia.org/wiki/Multivariate_normal_distribution),
and the realised niches are created by defining the positive and
negative interactions amongst a community of fundamental niches. A full
description of the `virtualNicheR` package, including connections to
niche theory and an explanation of the underlying mathematics, can be
found in the associated software paper.

## Citation

If you use `virtualNicheR` in your research we would be very grateful if
you could please cite the software using:

Etherington TR, Omondiagbe, OP (2019) virtualNicheR: creating virtual
fundamental and realised niches

## Installation

You can install the released version of `virtualNicheR` from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("virtualNicheR")
```

You can also install `virtualNicheR` from GitHub with:

``` r
# install.packages("githubinstall")
githubinstall::githubinstall("manaakiwhenua/virtualNicheR")

# install.packages("devtools")
devtools::install_github("manaakiwhenua/virtualNicheR")
```

## Dependencies

`virtualNicheR` is not dependant on any additional packages, and was
developed using `R` version 3.5.1. and is known to work for at least
`R` version 3.2.3.

## Quick Examples

In `virtualNicheR` the **fundamental niche** of a virtual species *S* is
defined by:

  - λ<sub>*max*</sub>, the species’ maximum finite rate of increase at
    the fundamental niche optimum,
  - **μ**, a vector of means that gives the optimum location of the
    niche in each dimension, and
  - **Σ**, a variance-covariance matrix that gives the size and
    orientation of the niche in each dimension.

By generating a systematic grid of coordinates in niche space, the niche
can be easily visualised.

``` r
library(virtualNicheR)

# Create coordinates across niche space
niche.XY = niche.grid.coords(mins=c(15,0), maxs=c(35,200), nCoords=121)

# Define a species as a function of the maximum finite rate of increase,
# a means vector, and covariance matrix
lambdaMax = 2.5
meansVector = matrix(c(25, 100))
covarMatrix = matrix(data=c(9, 60,
                            60, 625), nrow=2, ncol=2, byrow = TRUE)
species = list(lambdaMax, meansVector, covarMatrix)

# Calculate the fundamental niche
fundNiche = fund.niche(niche.XY, species)

# Plot the fundamental niche
fundNicheMatrix = matrix(fundNiche, nrow=length(unique(niche.XY[,1])))
nContour = 10
filled.contour(unique(niche.XY[,1]), unique(niche.XY[,2]), fundNicheMatrix,
               levels = seq(0, lambdaMax, lambdaMax/nContour),
               col=colorRampPalette(c("gold", "firebrick"))(nContour),
               xlab=expression(paste("Temperature (", degree, "C)")),
               ylab="Rainfall (mm)",
               main ="Fundamental niche",
               key.title = title(main = expression(lambda)))
```

<img src="man/figures/README-fund-niche-example-1.png" width="100%" />

Given maps of environmental data, the fundamental niche can also be
mapped in geographic space to create a **potential niche**.

``` r
# Map the potential niche given maps of environmental variables
# Convert matrices of variables into columns
temp1D = matrix(temperatureMap, ncol=1)
rain1D = matrix(rainfallMap, ncol=1)
data.XY = cbind(temp1D, rain1D)
# Calculate the potential niche and form back in a 2D map
poteNiche = fund.niche(data.XY, species)
poteNiche2D = matrix(poteNiche, ncol=100)
filled.contour(z=poteNiche2D,
              levels = seq(0, lambdaMax, lambdaMax/nContour),
              col=colorRampPalette(c("gold", "firebrick"))(nContour),
              asp=1, plot.axes = {}, frame.plot=FALSE,
              main ="Map of the potential niche",
              key.title = title(main = expression(lambda)))
```

<img src="man/figures/README-pote-niche-example-1.png" width="100%" />

Having created fundamental niches for several species, by defining a
matrix that quantifies the interactions (competition, predation,
commensalism, amensalism, or mutualism) between species, it is also
possible to calculate a **realised niche** that accounts for positive
and negative affects of species interactions.

``` r
# Define the community
species1 = list(2.5, matrix(c(25, 100)), matrix(data=c(9, 60,
                                                       60, 625), nrow=2, ncol=2, byrow=TRUE))
species2 = list(5.0, matrix(c(28, 110)), matrix(data=c(4, -20, 
                                                      -20, 500), nrow=2, ncol=2, byrow=TRUE))
species3 = list(3.0, matrix(c(25, 80)), matrix(data=c(4, 0,
                                                      0, 150), nrow=2, ncol=2, byrow=TRUE))
community = list(species1, species2, species3)

# Define the interactions
interactions = matrix(data=c(0.0, 0.2,-0.4,
                            -1.0, 0.0,-1.0,
                            -0.6, 0.3, 0.0), nrow=3, ncol=3,byrow = TRUE)

# Calculate the realised niche values
niche.XY = niche.grid.coords(mins=c(15,0), maxs=c(35,200), nCoords=121)
realNiche = real.niche(niche.XY, community, interactions)

# Plot the realised niche for species 1
realNiche1 = matrix(realNiche[,1], nrow=length(unique(niche.XY[,1])))
nContour = 10
filled.contour(unique(niche.XY[,1]), unique(niche.XY[,2]), realNiche1,
               levels = seq(0, 5, 5/nContour),
               col=colorRampPalette(c("gold", "firebrick"))(nContour),
               xlab=expression(paste("Temperature (", degree, "C)")),
               ylab="Rainfall (mm)",
               main ="Realised niche",
               key.title = title(main = expression(lambda)))
```

<img src="man/figures/README-real-niche-example-1.png" width="100%" />


NEWS.md
# virtualNicheR 0.1.0

1. First release. `fund.niche()` ,`real.niche`, and `niche.grid.coords`


tests/testhat.R
library(testthat)

test_check("virtualNicheR")


tests/testthat/test-fund-niche.R
context("virtualNicheR")


fund.niche.input<-function(){
  nCoords=5
  min=c(15,0)
  max=c(35,200)
  niche.XY  <- niche.grid.coords(mins=min, maxs=max, nCoords=nCoords)
  lambdaMax = 2.5
  meansVector = c(25, 100)
  covarMatrix = matrix(data=c(9, 60,
                              60, 625), nrow=2, ncol=2, byrow = TRUE)

  species = list(lambdaMax, meansVector, covarMatrix)
  fundNiche = fund.niche(niche.XY, species)
  result=list("fundN"=fundNiche,"nCoords"=nCoords,"max"=max,"min"=min,"lambdaMax"=lambdaMax,"meansVector"=meansVector,"covarMatrix"=covarMatrix)

  return(result)
}

test_that("Fundamental Niche result should always equal the number of coordinate raise to the power of the number of min/max value", {

  fundNiche<-fund.niche.input()
  expectedResult= fundNiche$nCoords ** length(fundNiche$min)


  expect_equal(expectedResult,length(fundNiche$fundN),info="The length of the fundamental niche result is not equal to the expected result")
})

test_that("Fundamental Niche produces values which are normally distributed", {

  fundNiche<-fund.niche.input()
  test.result=shapiro.test(fundNiche$fundN)

  expect_equal(if(0.1 > test.result$p.value)print(TRUE),TRUE,info="The fundamental niche result is not normally distributed")
})


test_that("Fundamental Niche multivariate normal distribution is correctly centered on the means", {
  fundNiche<-fund.niche.input()
  species = list(fundNiche$lambdaMax,  fundNiche$meansVector,  fundNiche$covarMatrix)
  fundNicheMeans = fund.niche(matrix(fundNiche$meansVector, ncol = ncol(fundNiche$covarMatrix)), species)
  expect_equal(if(all(fundNicheMeans >= fundNiche$fundN)) print(TRUE),TRUE,info="Fundamental Niche multivariate normal distribution is not correctly centered on the means")

})


test_that("fundamental niche functions always return non negative values",{
  fundNiche<-fund.niche.input()

  expect_equal(if(all(fundNiche$fundN >= 0)) print(TRUE),TRUE,info="The fundamental niche result is not unimodal")
})



test_that("Fundamental Niche produces vector of length equal to the number of row in the inputted niche space coordinates", {
  niche.XY  <- niche.grid.coords(mins=c(15,0), maxs=c(35,200), nCoords=5)
  lambdaMax = 2.5
  meansVector = c(25, 100)
  covarMatrix = matrix(data=c(9, 60,
                              60, 625), nrow=2, ncol=2, byrow = TRUE)
  species = list(lambdaMax, meansVector, covarMatrix)
  fundNiche = fund.niche(niche.XY, species)
  expect_equal(length(niche.XY [[2]]),length(fundNiche),info = "The fundamental niche length is not equal to the length of the grid niche space")
})


test_that("Fundamental Niche should not work
          # when the covariance matrix is not symmetric", {

  niche.XY  <- niche.grid.coords(mins=c(15,0), maxs=c(35,200), nCoords=5)
  lambdaMax = 2.5
  meansVector = c(25, 100)
  covarMatrix = matrix(data=c(195, 60,
                              60, 625,23,45), nrow=3, ncol=3, byrow = TRUE)
  species = list(lambdaMax, meansVector, covarMatrix)

  ## If covariance matrix is not symmetric, Fundamental Niche should return an error
  expect_error(fund.niche(niche.XY, species),"Covariance matrix is not symmetric")
})


test_that("Fundamental Niche should not work when
          # the length of meansVector is not equal to the number of rows of covarMatrix",{

  niche.XY  <- niche.grid.coords(mins=c(15,0), maxs=c(35,200), nCoords=5)
  lambdaMax = 2.5
  meansVector = c(25, 100,300)
  covarMatrix = matrix(data=c(195, 60,
                              60, 625), nrow=2, ncol=2, byrow = TRUE)
  species = list(lambdaMax, meansVector, covarMatrix)

  ## If Dimensions of means vector does not match covariance matrix, Fundamental Niche should return an error
  expect_error(fund.niche(niche.XY, species),"Dimensions of means vector does not match covariance matrix")
})


test_that("Fundamental Niche should not work when the  number of column
          # of the niche.coords is not equal to the length of meansVector", {

  niche.XY  <- niche.grid.coords(mins=c(15,5,25), maxs=c(35,200,400), nCoords=5)
  lambdaMax = 2.5
  meansVector = c(25, 100)
  covarMatrix = matrix(data=c(195, 60,
                              60, 625), nrow=2, ncol=2, byrow = TRUE)
  species = list(lambdaMax, meansVector, covarMatrix)

  ## If the number of column of the niche.coords not equal to length of meansVector, Fundamental Niche should return an error
  expect_error(fund.niche(niche.XY, species),"Dimensions of niche space does not match dimensions of defined niche")
})


tests/testthat/test-real-niche.R
context("virtualNicheR")


real.niche.input<-function(){
     nCoords=10
     min=c(15,0)
     max=c(35,200)
    species1 = list(2.5, c(25, 100), matrix(data=c(9, 60,60, 625), nrow=2, ncol=2, byrow=TRUE))
    species2 = list(5.0, c(28, 110), matrix(data=c(4, -20,
                                                  -20, 500), nrow=2, ncol=2, byrow=TRUE))
    species3 = list(3.0, c(25, 80), matrix(data=c(4, 0,
                                                  0, 150), nrow=2, ncol=2, byrow=TRUE))
    community = list(species1, species2, species3)

    # Define the interactions
    interactions = matrix(data=c(0.0, 0.2,-0.4,
                                -1.0, 0.0,-1.0,
                                -0.6, 0.3, 0.0), nrow=3, ncol=3, byrow = TRUE)

    # Calculate the realised niche values
    niche.XY = niche.grid.coords(mins=min, maxs=max, nCoords=nCoords)
    realNiche = real.niche(niche.XY, community, interactions)
    result=list("realN"=realNiche,"nCoords"=nCoords,"max"=max,"min"=min)
    return(result)
}


test_that("Real niche functions always return non negative values",{
  realNiche=real.niche.input()
  expect_equal(if(all(realNiche$realN >= 0)) print(TRUE),TRUE,info="The real niche result is not normally distributed")
})


test_that("Real Niche result should always equal the number of coordinate raise to the power of the number of min/max value", {

  realNiche<-real.niche.input()
  expectedResult= realNiche$nCoords ** length(realNiche$min)

  expect_equal(expectedResult,length(realNiche$realN[,1]),info="The length of the real niche result is not equal to the expected result")
})


test_that("Real Niche should not work when
          # the interactions is not a square matrix,", {
            species1 = list(2.5, c(25, 100), matrix(data=c(9, 60,
                                                           60, 625), nrow=2, ncol=2, byrow=TRUE))
            species2 = list(5.0, c(28, 110), matrix(data=c(4, -20,
                                                           -20, 500), nrow=2, ncol=2, byrow=TRUE))
            species3 = list(3.0, c(25, 80), matrix(data=c(4, 0,
                                                           0, 150), nrow=2, ncol=2, byrow=TRUE))
            community = list(species1, species2, species3)
            interactions = suppressWarnings(matrix(data=c(0.0, 0.2,-0.4,
                                           -1.0, 0.0,-1.0,
                                           -0.6, 0.3, 0.0,
                                           -0.2, 0.1, 0.2), ncol=3, byrow = TRUE))

            niche.XY = niche.grid.coords(mins=c(15,0), maxs=c(35,200), nCoords=121)


            ## If the interactions is not a square matrix, Real Niche should return an error
            expect_error(real.niche(niche.XY, community, interactions),"The interactions must be a square matrix")
})


test_that("Real Niche should not work when
          # the interactions matrix is not equal to zero on the diagonal,", {
            species1 = list(2.5, c(25, 100), matrix(data=c(9, 60,
                                                           60, 625), nrow=2, ncol=2, byrow=TRUE))
            species2 = list(5.0, c(28, 110), matrix(data=c(4, -20,
                                                           -20, 500), nrow=2, ncol=2, byrow=TRUE))
            species3 = list(3.0, c(25, 80), matrix(data=c(4, 0,
                                                          0, 150), nrow=2, ncol=2, byrow=TRUE))
            community = list(species1, species2, species3)

            interactions = matrix(data=c(0.0, 0.2,-0.4,
                                         -1.0, 0.0,-1.0,
                                         -0.6, 0.3, 0.0), nrow=3, ncol=3, byrow = TRUE)
            diag(interactions) <-2.0
            niche.XY = niche.grid.coords(mins=c(15,0), maxs=c(35,200), nCoords=121)

            ## If the interactions is not a square matrix, Real Niche should return an error
            expect_error(real.niche(niche.XY, community, interactions),"Interactions matrix must be equal to zero on the diagonal")
})


test_that("Real Niche should not work when
          # community size is different from the interactions size,", {
            species1 = list(2.5, c(25, 100), matrix(data=c(9, 60,
                                                           60, 625), nrow=2, ncol=2, byrow=TRUE))
            species2 = list(5.0, c(28, 110), matrix(data=c(4, -20,
                                                           -20, 500), nrow=2, ncol=2, byrow=TRUE))

            community = list(species1, species2)

            interactions = matrix(data=c(0.0, 0.2,-0.4,
                                         -1.0, 0.0,-1.0,
                                         -0.6, 0.3, 0.0), nrow=3, ncol=3, byrow = TRUE)
            niche.XY = niche.grid.coords(mins=c(15,0), maxs=c(35,200), nCoords=121)

            ## If the community size is not the same as the interactions size, Real Niche should return an error
            expect_error(real.niche(niche.XY, community, interactions),"The community size is not the same as the interactions size")
})



R/fund-niche.R
#-------------------------------------------------------------

#' @title Fundamental Niche

#' @description Calculates the \code{n}-dimensional fundamental niche at \code{m}
#' coordinates in niche space.
#'
#' @param niche.coords A \code{m}-by-\code{n} matrix of niche coordinates.
#' @param species Ordered list of: maximum finite rate of increase, a means vector, and
#' variance-covariance matrix that together define the fundamental niche of the
#' species.
#'
#' @details Fundamental niche values are calculated on the basis of a multi-variate
#' normal distribution, and hence are unimodal and convex in shape.  The
#' \code{niche.coords} can either be coordinates of specific interest, such as for
#' mapping niche suitability for a study area, or for visualisation in niche space a grid
#' of systematic coordinates as can be generated using
#' \code{\link{niche.grid.coords}}.  The \code{species} ordered list should contain:
#'
#' \itemize{
#'   \item maximum finite rate of increase, the species' maximum finite rate of
#'   increase at the fundamental niche optimum,
#'   \item a \code{n}-by-\code{1} column vector of means that gives the optimum location of the
#'   niche in each dimension, and
#'   \item a \code{n}-by-\code{n} variance-covariance matrix, that gives the size and orientation of the
#'   niche in each dimension.
#' }
#'
#' @return Returns a vector of length \code{m} containing the fundamental niche
#' value for each of the input niche space coordinates.
#'
#' @seealso Used internally by \code{\link{real.niche}}
#'
#' @references
#' Etherington TR, Omondiagbe, OP (2019) virtualNicheR: creating virtual fundamental and realised niches
#'
#' @examples
#' # Create coordinates across niche space
#' niche.XY = niche.grid.coords(mins=c(15,0), maxs=c(35,200), nCoords=121)
#' # Define a species as a function of the maximum finite rate of increase,
#' # a means vector, and covariance matrix
#' lambdaMax = 2.5
#' meansVector = matrix(c(25, 100))
#' covarMatrix = matrix(data=c(9, 60,
#'                             60, 625), nrow=2, ncol=2, byrow = TRUE)
#' species = list(lambdaMax, meansVector, covarMatrix)
#' # Calculate the fundamental niche
#' fundNiche = fund.niche(niche.XY, species)
#' # Plot the fundamental niche
#' fundNicheMatrix = matrix(fundNiche, nrow=length(unique(niche.XY[,1])))
#' nContour = 10
#' filled.contour(unique(niche.XY[,1]), unique(niche.XY[,2]), fundNicheMatrix,
#'                levels = seq(0, lambdaMax, lambdaMax/nContour),
#'                col=colorRampPalette(c("gold", "firebrick"))(nContour),
#'                xlab=expression(paste("Temperature (", degree, "C)")),
#'                ylab="Rainfall (mm)",
#'                main ="Fundamental niche",
#'                key.title = title(main = expression(lambda)))
#'
#' # Map the potential niche given maps of environmental variables
#' # Convert matrices of variables into columns
#' temp1D = matrix(temperatureMap, ncol=1)
#' rain1D = matrix(rainfallMap, ncol=1)
#' data.XY = cbind(temp1D, rain1D)
#' # Calculate the potential niche and form back in a 2D map
#' poteNiche = fund.niche(data.XY, species)
#' poteNiche2D = matrix(poteNiche, ncol=100)
#' filled.contour(z=poteNiche2D,
#'                levels = seq(0, lambdaMax, lambdaMax/nContour),
#'                col=colorRampPalette(c("gold", "firebrick"))(nContour),
#'                asp=1, plot.axes = {}, frame.plot=FALSE,
#'                main ="Map of the potential niche",
#'                key.title = title(main = expression(lambda)))
#' @importFrom stats mahalanobis
#' @export

fund.niche <- function(niche.coords, species) {

  lambdaMax = species[[1]]
  meansVector = species[[2]]
  covarMatrix = species[[3]]

  # Check the input data
  if (!isSymmetric(covarMatrix)) {
    stop("Covariance matrix is not symmetric")
  }
  if (length(meansVector) != nrow(covarMatrix)) {
    stop("Dimensions of means vector does not match covariance matrix")
  }
  if (ncol(niche.coords) != length(meansVector)) {
    stop("Dimensions of niche space does not match dimensions of defined niche")
  }

  # Calculate the fundamental niche
  return(lambdaMax * exp(-0.5 * mahalanobis(niche.coords, meansVector, covarMatrix)))

}

#-------------------------------------------------------------


R/niche-grid-coords.R
#-------------------------------------------------------------

#' @title Niche Grid Coordinates

#' @description Create an \code{n}-dimensional grid of coordinates across niche space.
#'
#' @param mins Vector of length \code{n} listing the niche space minimum for each
#' dimension.
#' @param maxs Vector of length \code{n} listing the niche space maximum for each
#' dimension.
#' @param nCoords Number of coordinates across the niche space in all dimensions.
#'
#' @details This function creates a grid of coordinates systematically located
#' throughout the specified niche space to enable visualisation of niche patterns.
#' The extent of the grid is given by the \code{mins} and \code{maxs}, and the number
#' of coordinates for each dimension is given by \code{nCoords}.
#'
#' @return A matrix with \code{n} columns.
#'
#' @examples
#' # Niche space grid coordinates usage
#' niche.XY = niche.grid.coords(mins=c(15,0), maxs=c(35,200), nCoords=5)

#' @export
niche.grid.coords <- function(mins, maxs, nCoords) {

  # Check input data
  if (length(mins) != length(maxs)) {
    stop("Length of mins and maxs differ")
  }
  if (FALSE %in% (mins < maxs)) {
    stop("Maximums not greater than minimums in all dimensions")
  }
  if (nCoords <= 1) {
    stop("Sample size must be greater than one")
  }

  # Create list of coordinate locations for each dimension
  dimCoords = list()
  for (n in seq(1, length(mins))) {
    dimCoords[[n]] = seq(mins[n], maxs[n], (maxs[n] - mins[n]) / (nCoords - 1))
  }
  # Create all combinations of coordinates across all dimension
  return(expand.grid(dimCoords))

}

#-------------------------------------------------------------


R/real-niche.R
#-------------------------------------------------------------

#' @title Realised Niche

#' @description Calculates the \code{n}-dimensional realised niche for a community
#' of \code{s} species at \code{m} coordinates in niche space.
#'
#' @param niche.coords A \code{m}-by-\code{n} matrix of niche coordinates.
#' @param community List of length \code{s} containing definitions of the
#' fundamental niches of species in the community.
#' @param interactions A \code{s}-by-\code{s} matrix that defines the \code{species}
#' interspecific interactions.
#'
#' @details The first step in calculating the realised niches is to calculate a
#' fundamental niche for each species in the \code{community} list.  This is done
#' using the \code{\link{fund.niche}} function, so see there for a full description
#' of how to define a the species fundamental niches.
#'
#' The \code{interactions} matrix defines the form of the ecological interaction
#' between each of the species in the community.  These interactions can be either
#' positive (+) or negative (-) in effect, and as the interactions can be
#' asymmetrical there are a range of possible interspecific species interactions:
#' competition (--), predator-prey or parasite-host (+-), commensalism (+0),
#' amensalism (-0), or mutualism (++).  When building the interactions matrix
#' row 1 column 2 is the effects of species 2 on species 1, and conversely row 2
#' column 1 is the effects of species 1 on species 2, etc.
#'
#' Please note that this function assumes that the species are consistently
#' ordered in the community list and interactions matrix, and that no species effects
#' itself so the diagonal of the interactions matrix has values equal to zero.
#'
#' @return Returns a \code{m}-by-\code{s} matrix realised niche value for each species at
#' each of the input niche space coordinates.
#'
#' @seealso Uses \code{\link{fund.niche}}
#'
#' @references
#' Etherington TR, Omondiagbe, OP (2019) virtualNicheR: creating virtual fundamental and realised niches
#'
#' @examples
#' # Define the community
#' species1 = list(2.5, matrix(c(25, 100)), matrix(data=c(9, 60,
#'                                                        60, 625), nrow=2, ncol=2, byrow=TRUE))
#' species2 = list(5.0, matrix(c(28, 110)), matrix(data=c(4, -20,
#'                                                       -20, 500), nrow=2, ncol=2, byrow=TRUE))
#' species3 = list(3.0, matrix(c(25, 80)), matrix(data=c(4, 0,
#'                                                       0, 150), nrow=2, ncol=2, byrow=TRUE))
#' community = list(species1, species2, species3)
#'
#' # Define the interactions
#' interactions = matrix(data=c(0.0, 0.2,-0.4,
#'                             -1.0, 0.0,-1.0,
#'                             -0.6, 0.3, 0.0), nrow=3, ncol=3, byrow = TRUE)
#'
#' # Calculate the realised niche values
#' niche.XY = niche.grid.coords(mins=c(15,0), maxs=c(35,200), nCoords=121)
#' realNiche = real.niche(niche.XY, community, interactions)
#'
#' # Plot the realised niche for species 1
#' realNiche1 = matrix(realNiche[,1], nrow=length(unique(niche.XY[,1])))
#' nContour = 10
#' filled.contour(unique(niche.XY[,1]), unique(niche.XY[,2]), realNiche1,
#'                levels = seq(0, 5, 5/nContour),
#'                col=colorRampPalette(c("gold", "firebrick"))(nContour),
#'                xlab=expression(paste("Temperature (", degree, "C)")),
#'                ylab="Rainfall (mm)",
#'                main ="Realised niche",
#'                key.title = title(main = expression(lambda)))
#'
#' @export
real.niche <- function(niche.coords, community, interactions) {

  # Check the input data
  if (ncol(interactions) != nrow(interactions)) {
    stop("The interactions must be a square matrix")
  }
  if (FALSE %in% (diag(interactions) == 0)) {
    stop("Interactions matrix must be equal to zero on the diagonal")
  }
  if (length(community) != nrow(interactions)) {
    stop("The community size is not the same as the interactions size")
  }

  # Calculate fundamental niche for each species and convert to matrix
  fundNiches = c()
  for (species in community) {
    fundNiches = c(fundNiches, fund.niche(niche.coords, species))
  }
  fundNiches = matrix(fundNiches, ncol=length(community))

  # Calculate the realised niches
  realNiches = fundNiches * exp(t(apply(fundNiches, MARGIN=1, FUN="%*%", t(interactions))))
  return(realNiches)

}

#-------------------------------------------------------------
